src/common/convenience.h

```diff
@@ -0,0 +1,164 @@
+#include <type_traits>
+#include <utility>
+
+#include <boost/thread/shared_mutex.hpp>
+
+#include "common/backport14.h"
+#include "common/shunique_lock.h"
+
+#include "include/assert.h" // I despise you. Not you the reader, I'm talking
+                            // to the include file.
+
+
+#ifndef CEPH_COMMON_CONVENIENCE_H
+#define CEPH_COMMON_CONVENIENCE_H
+
+namespace ceph {
+
+// Lock Factories
+// ==============
+//
+// I used to, whenever I declared a mutex member variable of a class,
+// declare a pile of types like:
+// ```cpp
+// using unique_lock = ::std::unique_lock<decltype(membermutex)>;
+// ```
+// to avoid having to type that big, long type at every use. It also
+// let me change the mutex type later. It's inelegant and breaks down
+// if you have more than one type of mutex in the same class. So here
+// are some lock factories.
+template<typename Mutex, typename ...Args>
+inline auto uniquely_lock(Mutex&& m, Args&& ...args)
+  -> std::unique_lock<remove_reference_t<Mutex> > {
+  return std::unique_lock<remove_reference_t<Mutex> >(
+    std::forward<Mutex>(m), std::forward<Args>(args)... );
+}
+
+template<typename Mutex, typename ...Args>
+inline auto sharingly_lock(Mutex&& m, Args&& ...args)
+  -> boost::shared_lock<remove_reference_t<Mutex> > {
+  return
+    boost::shared_lock<remove_reference_t<Mutex> >(
+      std::forward<Mutex>(m), std::forward<Args>(args)...);
+}
+
+template<typename Mutex, typename ...Args>
+inline auto shuniquely_lock(std::unique_lock<Mutex>&& m, Args&& ...args)
+  -> shunique_lock<remove_reference_t<Mutex> > {
+  return shunique_lock<remove_reference_t<Mutex> >(
+    std::forward<std::unique_lock<Mutex> >(m), std::forward<Args>(args)...);
+}
+
+template<typename Mutex, typename ...Args>
+inline auto shuniquely_lock(boost::shared_lock<Mutex>&& m, Args&& ...args)
+  -> shunique_lock<remove_reference_t<Mutex> > {
+  return shunique_lock<remove_reference_t<Mutex> >(
+    std::forward<boost::shared_lock<Mutex> >(m),
+    std::forward<Args>(args)...);
+}
+
+template<typename Mutex, typename ...Args>
+inline auto shuniquely_lock(Mutex&& m, Args&& ...args)
+  -> shunique_lock<remove_reference_t<Mutex> > {
+  return shunique_lock<remove_reference_t<Mutex> >(
+    std::forward<Mutex>(m), std::forward<Args>(args)...);
+}
+
+// All right! These two don't work like the others. You cannot do
+// `auto l = guardedly_lock(m)` since copy elision before C++17 is
+// optional. C++17 makes it mandatory so these workarounds won't be
+// needed.
+//
+// To use this, you'll need to do something like:
+// `auto&& l = guardedly_lock(m)`
+// This way, we aren't actually copying or moving a value, we're
+// binding a reference to a temporary which extends its lifetime until
+// the end of the enclosing block.
+//
+// You may in fact want to use
+// `[[gnu::unused]] auto&& l = guardedly_lock(m)`
+// To avoid the unused variable warning. Since reference assignment
+// doesn't have side effects, normally, this just looks to the
```

Comment:
I like coding guidelines as much as the next person, but banishing contractions from documentation is a bit ridiculous, don't you think?